plasmapy/particles/tests/test_ionization_state.py

```diff
@@ -1,19 +1,100 @@
-import pytest
 import collections
 
-import numpy as np
 import astropy.units as u
+import numpy as np
+import pytest
 
-from plasmapy.particles.ionization_state import IonizationState
-from plasmapy.utils.pytest_helpers import run_test
-from plasmapy.particles.exceptions import AtomicError, InvalidIsotopeError
 from plasmapy.particles import (
+    Particle,
     atomic_number,
     atomic_symbol,
-    particle_symbol,
     isotope_symbol,
-    Particle,
+    particle_symbol,
 )
+from plasmapy.particles.exceptions import InvalidIsotopeError, ParticleError
+from plasmapy.particles.ionization_state import IonicFraction, IonizationState
+from plasmapy.utils.pytest_helpers import run_test
+
+ionic_fraction_table = [
+    ("Fe 6+", 0.52, 5.2e-6 * u.m ** -3),
+    ("He 1+", None, None),
+    ("H-2 0+", None, None),
+]
+
+
+@pytest.mark.parametrize("ion, ionic_fraction, number_density", ionic_fraction_table)
+def test_ionic_fraction_attributes(ion, ionic_fraction, number_density):
+
+    try:
+        instance = IonicFraction(
+            ion=ion, ionic_fraction=ionic_fraction, number_density=number_density
+        )
+    except Exception:
+        pytest.fail(
+            f"Unable to instantiate IonicFraction for ion = {ion}, "
+            f"ionic_fraction = {ionic_fraction}, and number_density = "
+            f"{number_density}"
+        )
+
+    # Prepare to check for the default values when they are not set
+
+    if ionic_fraction is None:
+        ionic_fraction = np.nan
+    if number_density is None:
+        number_density = np.nan * u.m ** -3
+
+    assert Particle(ion) == Particle(instance.ionic_symbol)
+    assert u.isclose(instance.ionic_fraction, ionic_fraction, equal_nan=True)
+    assert u.isclose(instance.number_density, number_density, equal_nan=True)
+
+
+@pytest.mark.parametrize(
+    "invalid_fraction, expected_exception",
+    [(-1e-9, ValueError), (1.00000000001, ValueError), ("...", TypeError),],
+)
+def test_ionic_fraction_invalid_inputs(invalid_fraction, expected_exception):
+    """
+    Test that IonicFraction raises exceptions when the ionic fraction
+    is out of the interval [0,1] or otherwise invalid.
+    """
+    with pytest.raises(expected_exception):
+        IonicFraction(ion="Fe 6+", ionic_fraction=invalid_fraction)
+        pytest.fail(
+            f"IonicFraction did not raise the appropriate exception "
+            f"for ionic_fraction = {repr(invalid_fraction)}, which is "
+            f"invalid."
+        )
+
+
+@pytest.mark.parametrize("invalid_particle", ["H", "e-", "Fe-56"])
+def test_ionic_fraction_invalid_particles(invalid_particle):
+    """
+    Test that `~plasmapy.particles.IonicFraction` raises the appropriate
+    exception when passed a particle that isn't a neutral or ion.
+    """
+    with pytest.raises(ParticleError):
+        IonicFraction(invalid_particle, ionic_fraction=0)
+        pytest.fail()
```

Comment:
um, what's this `pytest.fail` doing here?

Reply:
Removed.